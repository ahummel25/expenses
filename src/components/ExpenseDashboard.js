import React from 'react';
import ExpenseList from './ExpenseList';

const ExpenseDashboard = () => (
    <div>
     <ExpenseList />
    </div>
);

export default ExpenseDashboard;