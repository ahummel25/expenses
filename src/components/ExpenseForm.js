import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

const now = moment().format('MMM Do, YYYY');

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };

    this.onDescriptionChange = this.onDescriptionChange.bind(this);
    this.onNoteChange = this.onNoteChange.bind(this);
    this.onAmountChange = this.onAmountChange.bind(this);
    this.onDateChange = this.onDateChange.bind(this);
    this.onFocusChange = this.onFocusChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }  

onDescriptionChange = (event) => {
   const description = event.target.value;
   this.setState(() => ({ description }));
};

onNoteChange = (event) => {
    const note = event.target.value;
    this.setState(() => ({ note }));
};

onAmountChange = (event) => {
    const amount = event.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }  
};

onDateChange = (createdAt) => {
  if (createdAt) {
    this.setState(() => ({ createdAt }));
  }
};

onFocusChange = ({ focused }) => {
  this.setState(() => ({ calendarFocused: focused }));
};

onSubmit = (event) => {
  event.preventDefault();

  if (!this.state.description || !this.state.amount) {
    const error = 'Please provide description and amount.';
    this.setState(() => ({ error }));
  } else {
    this.setState(() => ({ error: '' }));
    this.props.onSubmit({
      description: this.state.description,
      amount: parseFloat(this.state.amount, 10) * 100,
      createdAt: this.state.createdAt.valueOf(),
      note: this.state.note
    })
  }
};

  render() {

    const errorStyle = {
      color: 'red',
      fontStyle: 'italic'
    };

    return (
      <div>
       {this.state.error && <p style={errorStyle}>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input 
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input 
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea 
            placeholder="Add a note for your expense"
            value={this.state.note}
            onChange={this.onNoteChange}
            >
          </textarea>
          <button>Add expense</button>
        </form>     
      </div>
    )
  }
}

export default ExpenseForm;