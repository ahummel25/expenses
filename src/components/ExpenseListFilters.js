import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter } from '../actions/filters';
import { sortByDate } from '../actions/filters';
import { sortByAmount } from '../actions/filters';

const ExpenseListFilters = (props) => (
  <div>
    <input type="text" value={props.filters.text} onChange={(event) => {
        props.dispatch(setTextFilter(event.target.value));
    }} />
    <select 
    value={props.filters.sortBy}
    onChange={(event) => {
       callFilterFunction(event.target.value, props.dispatch);
    }}>
      <option value="date">Date</option>
      <option value="amount">Amount</option>
    </select>
  </div>
);

const callFilterFunction = (filterValue, dispatch) => {
  if (filterValue === 'date') {
    dispatch(sortByDate());
  } else if (filterValue === 'amount') {
    dispatch(sortByAmount());
  }
};

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  }
};

export default connect(mapStateToProps)(ExpenseListFilters);