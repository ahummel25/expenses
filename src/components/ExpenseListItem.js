import React from 'react';

export const ExpenseListItem = ({ description, amount, note, createdAt }) => (
  <div>
    <h3>{description}</h3>
    <p>${amount} - {createdAt}</p>
  </div>
);