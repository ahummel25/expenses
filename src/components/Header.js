import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
    <header>
      <h1>Expensify</h1>
      <NavLink exact to="/" activeClassName="is-active">Home</NavLink>
      <NavLink to="/create" activeClassName="is-active">Create Expenses</NavLink>
    </header>
);

export default Header;