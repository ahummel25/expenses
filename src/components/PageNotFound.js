import React from 'react';
import { Link } from 'react-router-dom';

const PageNotFound = () => (
    <div>
     <p>404! Page Not Found. <Link to="/">Go Home</Link></p>
    </div>
);

export default PageNotFound;