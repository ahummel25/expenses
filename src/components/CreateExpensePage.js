import React from 'react';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export const CreateExpensePage = (props) => (
    <div>
     <h1>Add Expense</h1>
     <ExpenseForm 
       onSubmit={(expense) => {
          props.addExpense(expense);
          props.history.push('/');
       }}
     />
    </div>
);

const mapDispatchToProps = (dispatch) => ({ addExpense: (expense) => dispatch(addExpense(expense)) });

export default connect(undefined, mapDispatchToProps)(CreateExpensePage);