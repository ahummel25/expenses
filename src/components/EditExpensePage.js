import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
    this.onClick = this.onClick.bind(this);
  }

  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onClick = () => {
    const id = this.props.expense.id;
    this.props.removeExpense({ id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
      <ExpenseForm
        expense={this.props.expense}
        onSubmit={this.onSubmit}
      />
      <button onClick={this.onClick}>Remove</button>
    </div>
    )
  }
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (id) => dispatch(removeExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);