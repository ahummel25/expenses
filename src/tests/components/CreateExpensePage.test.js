import React from 'react';
import { shallow } from 'enzyme';
import { CreateExpensePage } from '../../components/CreateExpensePage';
import { expenses } from '../fixtures/expenses';

describe('test the CreateExpensePage', () => {
  let wrapper, addExpenseSpy, history;

  beforeEach(() => {
    addExpenseSpy = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<CreateExpensePage addExpense={addExpenseSpy} history={history}/>);
    expect(wrapper).toMatchSnapshot();
  });

  afterEach(() => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should handle onSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(addExpenseSpy).toHaveBeenLastCalledWith(expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
  });

});