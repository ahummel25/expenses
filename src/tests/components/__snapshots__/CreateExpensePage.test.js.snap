// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test the CreateExpensePage should handle onSubmit 1`] = `
<div>
  <h1>
    Add Expense
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;

exports[`test the CreateExpensePage should handle onSubmit 2`] = `
<div>
  <h1>
    Add Expense
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
